File: CourseSchedule2.py
class Solution:
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        def dfs_cycle(course):
            if visited[course] == 1:
                return False
            if visited[course] == -1:
                return True
            visited[course] = 1
            for i in graph[course]:
                if visited[i] == 1:
                    return False
                elif not dfs_cycle(i):
                    return False
            visited[course] = -1
            return True
            
        graph = [ [] for _ in range(numCourses)]
        for pre in prerequisites:
            graph[pre[1]].append(pre[0])
        visited = [0] * numCourses
        
        for i in range(numCourses):
            if not dfs_cycle(i):
                return False
        return True
        
if __name__ == '__main__':
    numCourses = int(raw_input())
    n = int(raw_input())
    prerequisites = []
    for i in range(n):
        pre = list(input())
        prerequisites.append(pre)
    solution = Solution()
    print(solution.canFinish(numCourses, prerequisites))

File: CourseSchedule2.cpp
/*
 * Time Limit Exceeded
 */
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int> >& prerequisites) {
        vector<vector<int> > graph(numCourses, vector<int>(numCourses, 0));
        vector<bool> visited(numCourses, false);
        
        for(int i=0;i<prerequisites.size();i++){
            graph[prerequisites[i].second][prerequisites[i].first] = 1;
        }
        
        for(int i=0;i<numCourses;i++){
            visited[i] = true;
            if(!dfs(i, numCourses, graph, visited)) return false;
        }
        return true;
        
    }
    
    bool dfs(int course, int n, vector<vector<int> >& graph, vector<bool>& visited){
        for(int i=0;i<n;i++){
            if(graph[course][i] == 1){
                if(visited[i]) return false;
                else{
                    visited[i] = true;
                    if(!dfs(i, n, graph, visited)) return false;
                }
            }
        }
        visited[course] = false;
        return true;
    }
};

int main(){
    
    int numCourses;
    cin>>numCourses;
    
    int n;
    cin>>n;
    vector<pair<int, int> > prerequisites;
    for(int i=0;i<n;i++){
        int a, b;
        cin>>a>>b;
        prerequisites.push_back(pair<int, int>(a, b));
    }
    
    Solution *solution = new Solution();
    solution->canFinish(numCourses, prerequisites) ? cout<<"True" : cout<<"False";
    
    return 0;
    
}
